### PR TITLE
feat(adminPanel): RN-1188: Edit survey on a new page

### PR DESCRIPTION
**Pull request by alexd-bes**
### Changes:
- Added ability to link edit button to a new url
- Added Edit Survey page, and associated refactoring to make components reusable
- Added `required` `*` to fields
- Updated `FileUploadField`
- Added server-side validation that selected countries exist in project
- Added ability to have multiple nested views in a page ( because now there are two views nested beneath `surveys` - details and edit) 

---

### Screenshots:
Screen Shot 2024-05-13 at 12 03 50 PM

### CODE REVIEW

**Review by jaskfla — COMMENTED** [review]
I started this review so followed through; but might flag this one for @rohan-bes to do a more in-depth review on—particularly bits involving Redux

Otherwise not a lot from me. Left two small questions where I don’t understand the code, but the changes are looking awesome! Impressively fast turnaround, though perhaps unsurprising by now 😆👏

File: packages/central-server/package.json [hunk 1]
```diff
@@ -22,7 +22,7 @@
     "start-dev": "yarn package:start:backend-start-dev 9999",
     "start-verbose": "LOG_LEVEL=debug yarn start-dev",
     "setup-test": "yarn workspace @tupaia/database setup-test-database",
-    "test": "yarn workspace @tupaia/database check-test-database-exists && NODE_ENV=test mocha",
+    "test": "yarn workspace @tupaia/database check-test-database-exists && DB_NAME='tupaia_test' NODE_ENV=test mocha",
```
[the remark this answers is not in the document]
Comment (alexd-bes) [reply packages/central-server/package.json:25]: Oh yes, this was accidental as it wasn't working for me locally

File: packages/central-server/src/tests/apiV2/surveys/CreateAndEditSurveys.test.js [hunk 2]
```diff
@@ -186,6 +226,54 @@ describe('Create and Edit Surveys', () => {
...
+
+        // Revert the countryIds back to Vanuatu for other test cases
+        await models.survey.update(
+          { code: EXISTING_TEST_SURVEY_CODE_1 },
+          { country_ids: [vanuatuCountry.id] },
+        );
+
+        expectResponseError(
+          response,
+          /The following countries are not part of the project: Tonga/,
```
[the remark this answers is not in the document]
Comment (alexd-bes) [reply packages/central-server/src/tests/apiV2/surveys/CreateAndEditSurveys.test.js:248]: In some cases (e.g. when creating a survey) there are some extra bits in the error message, so this makes sure that it matches every time

File: packages/lesmis/src/views/AdminPanel/routes/surveys/getSurveysPageConfig.js [hunk 3]
```diff
@@ -424,10 +424,12 @@ export const getSurveysPageConfig = translate => {
...
     deleteConfig: getDeleteConfigs(translate),
     editorConfig: getBaseEditorConfigs(translate),
-    nestedView: {
-      ...surveys.nestedView,
-      title: translate('admin.questions'),
-      columns: QUESTION_COLUMNS,
-    },
+    nestedViews: [
+      {
+        ...surveys.nestedViews[0],
```
[the remark this answers is not in the document]
Comment (alexd-bes) [reply packages/lesmis/src/views/AdminPanel/routes/surveys/getSurveysPageConfig.js:429]: `nestedViews` will always be an array of objects, so this shouldn't ever happen. With LESMIS, what happens is we need to return a copy of it with the translated titles etc. We are also keeping the edit modal for now, so we don't need the edit nested view

**Review by rohan-bes — APPROVED** [review]
Awesome stuff @alexd-bes, the look n feel of this is great! Happy with the code too!

Noticed a couple of little issues playing with the branch:
1. If I change the data-service from DHIS2 to Tupaia and try and save I get a 'data service must match' error. That itself isn't an issue (or at least I don't think it's caused by these changes) but after that any other survey I click on to edit will have 'Tupaia' as the data service, so I think there's some kind of redux state issue going on.
2. When saving a survey, on a larger monitor I can still see a little bit of the editor fields underneath the loading spinner:
image

Just these two issues tho, so pre-approving. Happy to give this another look over if ya want me to tho if either issue requires significant changes :)

**Comment by alexd-bes** [comment]
> Awesome stuff @alexd-bes, the look n feel of this is great! Happy with the code too!
> 
> Noticed a couple of little issues playing with the branch:
> 
> 1. If I change the data-service from DHIS2 to Tupaia and try and save I get a 'data service must match' error. That itself isn't an issue (or at least I don't think it's caused by these changes) but after that any other survey I click on to edit will have 'Tupaia' as the data service, so I think there's some kind of redux state issue going on.
> 2. When saving a survey, on a larger monitor I can still see a little bit of the editor fields underneath the loading spinner:
>    image
> 
> Just these two issues tho, so pre-approving. Happy to give this another look over if ya want me to tho if either issue requires significant changes :)

Thanks @rohan-bes ! I fixed the redux state issue, but couldn't replicate the loading issue - I tried on Firefox and Chrome at extra large screen sizes. If it still happens for you, could you give me the screen size you're working with, and any other useful info (browser etc)?

**Comment by rohan-bes** [comment]
> Thanks @rohan-bes ! I fixed the redux state issue, but couldn't replicate the loading issue - I tried on Firefox and Chrome at extra large screen sizes. If it still happens for you, could you give me the screen size you're working with, and any other useful info (browser etc)?

I tried to reproduce this morning and it was happening for a bit, but then after I encountered an error message it stopped happening even after refreshing page!? Not sure what's going on but happy to leave it and see if it shows up during testing :+1:

**Pull request merged**